File: app/src/main/java/org/bepass/oblivion/ui/MainActivity.java
package org.bepass.oblivion.ui;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.activity.OnBackPressedCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;

import org.bepass.oblivion.R;
import org.bepass.oblivion.base.StateAwareBaseActivity;
import org.bepass.oblivion.databinding.ActivityMainBinding;
import org.bepass.oblivion.enums.ConnectionState;
import org.bepass.oblivion.service.OblivionVpnService;
import org.bepass.oblivion.utils.FileManager;
import org.bepass.oblivion.utils.LocaleHandler;
import org.bepass.oblivion.utils.NetworkUtils;
import org.bepass.oblivion.utils.PublicIPUtils;
import org.bepass.oblivion.utils.ThemeHelper;

import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends StateAwareBaseActivity<ActivityMainBinding> {

    private static final String TAG = "MainActivity";
    private long backPressedTime;
    private Toast backToast;
    private LocaleHandler localeHandler;
    private ActivityResultLauncher<Intent> vpnPermissionLauncher;
    
    // Executor for background tasks to avoid blocking Main Thread
    private final ExecutorService backgroundExecutor = Executors.newSingleThreadExecutor();

    /**
     * Static helper to start the Activity
     */
    public static void start(Context context) {
        Intent starter = new Intent(context, MainActivity.class);
        starter.putExtra("origin", context.getClass().getSimpleName());
        starter.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        context.startActivity(starter);
    }

    /**
     * Starts the Foreground Service with all necessary configuration extras.
     */
    public static void startVpnService(Context context, Intent intent) {
        // Fetch settings efficiently
        boolean proxyMode = FileManager.getBoolean("USERSETTING_proxymode");
        String license = FileManager.getString("USERSETTING_license");
        int endpointType = FileManager.getInt("USERSETTING_endpoint_type");
        boolean gool = FileManager.getBoolean("USERSETTING_gool");
        String endpoint = FileManager.getString("USERSETTING_endpoint");
        String port = FileManager.getString("USERSETTING_port");
        boolean lan = FileManager.getBoolean("USERSETTING_lan");

        intent.putExtra("USERSETTING_proxymode", proxyMode);
        intent.putExtra("USERSETTING_license", license);
        intent.putExtra("USERSETTING_endpoint_type", endpointType);
        intent.putExtra("USERSETTING_gool", gool);
        intent.putExtra("USERSETTING_endpoint", endpoint);
        intent.putExtra("USERSETTING_port", port);
        intent.putExtra("USERSETTING_lan", lan);

        intent.setAction(OblivionVpnService.FLAG_VPN_START);
        ContextCompat.startForegroundService(context, intent);
    }

    @Override
    protected int getLayoutResourceId() {
        return R.layout.activity_main;
    }

    @Override
    protected int getStatusBarColor() {
        return R.color.status_bar_color;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Initialize Utilities
        localeHandler = new LocaleHandler(this);
        ThemeHelper.getInstance().updateActivityBackground(binding.getRoot());

        // Perform heavy file operations in background
        backgroundExecutor.execute(() -> FileManager.cleanOrMigrateSettings(getApplicationContext()));

        setupUI();
        setupVPNConnection();
        requestNotificationPermission();
        handleBackPress();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (backgroundExecutor != null && !backgroundExecutor.isShutdown()) {
            backgroundExecutor.shutdown();
        }
    }

    private void setupVPNConnection() {
        vpnPermissionLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(), result -> {
                    if (result.getResultCode() == RESULT_OK) {
                        handleVpnButton(true);
                    } else {
                        Toast.makeText(this, R.string.permission_denied_vpn, Toast.LENGTH_LONG).show();
                        updateConnectButtonState(false);
                    }
                });

        binding.connectionButton.setOnClickListener(v -> {
            boolean isConnected = !lastKnownConnectionState.isDisconnected();
            handleVpnButton(!isConnected);
        });
    }

    /**
     * Updates the connection button state (icon).
     */
    private void updateConnectButtonState(boolean isConnected) {
        if (isConnected) {
            binding.connectionIcon.setImageResource(R.drawable.vpn_on);
        } else {
            binding.connectionIcon.setImageResource(R.drawable.vpn_off);
        }
    }

    private void handleVpnButton(boolean enableVpn) {
        // Initialize FileManager if not already done (failsafe)
        FileManager.initialize(getApplicationContext());

        if (enableVpn) {
            if (lastKnownConnectionState.isDisconnected()) {
                Intent vpnIntent = OblivionVpnService.prepare(this);
                if (vpnIntent != null) {
                    vpnPermissionLauncher.launch(vpnIntent);
                } else {
                    vpnIntent = new Intent(this, OblivionVpnService.class);
                    startVpnService(this, vpnIntent);
                }
                // Basic network monitoring
                NetworkUtils.monitorInternetConnection(lastKnownConnectionState, this);
            } else if (lastKnownConnectionState.isConnecting()) {
                OblivionVpnService.stopVpnService(this);
            }
        } else {
            if (!lastKnownConnectionState.isDisconnected()) {
                OblivionVpnService.stopVpnService(this);
            }
        }
    }

    private void handleBackPress() {
        getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                if (backPressedTime + 2000 > System.currentTimeMillis()) {
                    if (backToast != null) backToast.cancel();
                    finish();
                } else {
                    if (backToast != null) backToast.cancel();
                    backToast = Toast.makeText(MainActivity.this, R.string.press_back_again, Toast.LENGTH_SHORT);
                    backToast.show();
                }
                backPressedTime = System.currentTimeMillis();
            }
        });
    }

    private void requestNotificationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            ActivityResultLauncher<String> pushNotificationPermissionLauncher = registerForActivityResult(
                    new ActivityResultContracts.RequestPermission(), isGranted -> {
                        if (!isGranted) {
                            // Optionally show a rationale dialog here
                            Log.w(TAG, "Notification permission denied");
                        }
                    });
            pushNotificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS);
        }
    }

    private void setupUI() {
        binding.floatingActionButton.setOnClickListener(v -> localeHandler.showLanguageSelectionDialog());
        binding.infoIcon.setOnClickListener(v -> startActivity(new Intent(MainActivity.this, InfoActivity.class)));
        binding.bugIcon.setOnClickListener(v -> startActivity(new Intent(MainActivity.this, LogActivity.class)));
        binding.settingIcon.setOnClickListener(v -> startActivity(new Intent(MainActivity.this, SettingsActivity.class)));
    }

    @NonNull
    @Override
    public String getKey() {
        return "mainActivity";
    }

    @Override
    protected void onResume() {
        super.onResume();
        observeConnectionStatus();
        // Refresh IP visibility in case connection dropped while paused
        if (lastKnownConnectionState.isDisconnected()) {
            binding.publicIP.setVisibility(View.GONE);
            binding.ipProgressBar.setVisibility(View.GONE);
        }
    }

    @Override
    public void onConnectionStateChange(ConnectionState state) {
        runOnUiThread(() -> {
            Log.d(TAG, "Connection state changed to: " + state);
            updateUI(state);
        });
    }

    /**
     * Centralized UI update logic based on connection state.
     */
    private void updateUI(ConnectionState state) {
        switch (state) {
            case DISCONNECTED:
                binding.publicIP.setVisibility(View.GONE);
                binding.stateText.setText(R.string.notConnected);
                binding.ipProgressBar.setVisibility(View.GONE);
                binding.connectionButton.setEnabled(true);
                updateConnectButtonState(false);
                break;

            case CONNECTING:
                binding.stateText.setText(R.string.connecting);
                binding.publicIP.setVisibility(View.GONE);
                binding.ipProgressBar.setVisibility(View.VISIBLE);
                binding.connectionButton.setEnabled(true);
                updateConnectButtonState(true);
                break;

            case CONNECTED:
                handleConnectedStateUI();
                break;
        }
    }

    private void handleConnectedStateUI() {
        binding.connectionButton.setEnabled(true);
        updateConnectButtonState(true);
        binding.ipProgressBar.setVisibility(View.GONE);

        // Handle Text Description
        boolean isProxy = FileManager.getBoolean("USERSETTING_proxymode");
        String port = FileManager.getString("USERSETTING_port");
        
        if (isProxy) {
            boolean isLan = FileManager.getBoolean("USERSETTING_lan");
            String ip = "127.0.0.1";
            String modeText = "socks5";
            
            if (isLan) {
                try {
                    ip = NetworkUtils.getLocalIpAddress(this);
                    modeText = "socks5 over LAN";
                } catch (Exception e) {
                    ip = "0.0.0.0";
                }
            }
            
            binding.stateText.setText(String.format(Locale.getDefault(), 
                    "%s\n%s on %s:%s", 
                    getString(R.string.connected), modeText, ip, port));
        } else {
            binding.stateText.setText(R.string.connected);
        }

        // Fetch and Show Public IP
        fetchPublicIP();
    }

    private void fetchPublicIP() {
        // binding.ipProgressBar.setVisibility(View.VISIBLE); // Removed to prevent confusion with connection loading
        PublicIPUtils.getInstance().getIPDetails(details -> runOnUiThread(() -> {
            // Lifecycle check to prevent crash
            if (isFinishing() || isDestroyed()) return;

            binding.ipProgressBar.setVisibility(View.GONE); // Ensure it's gone when done
            if (details.ip != null) {
                String ipInfo = details.ip + " " + (details.flag != null ? details.flag : "");
                binding.publicIP.setText(ipInfo);
                binding.publicIP.setVisibility(View.VISIBLE);
            } else {
                // Optional: Handle error state or leave hidden
                binding.publicIP.setVisibility(View.INVISIBLE);
            }
        }));
    }
}

File: app/src/main/java/org/bepass/oblivion/ui/MainActivity.kt
package org.bepass.oblivion.ui

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color as AndroidColor
import android.net.VpnService
import android.os.Build
import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.SystemBarStyle
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.core.content.ContextCompat
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import org.bepass.oblivion.R
import org.bepass.oblivion.enums.VpnCoreType
import org.bepass.oblivion.platform.VpnServiceConnector
import org.bepass.oblivion.service.OblivionVpnService
import org.bepass.oblivion.ui.theme.OblivionTheme
import org.bepass.oblivion.utils.FileManager
import org.bepass.oblivion.utils.FontSizeHelper
import org.bepass.oblivion.utils.LocaleManager
import org.bepass.oblivion.utils.ThemeHelper

@AndroidEntryPoint
class MainActivity : ComponentActivity() {
  @Inject lateinit var vpnServiceConnector: VpnServiceConnector

  private var pendingVpnStart = false
  private var pendingNotificationPermissionForVpnStart = false

  private val vpnPermissionLauncher =
    registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
      if (result.resultCode == RESULT_OK && pendingVpnStart) {
        startVpnServiceFromActivity(Intent(this, OblivionVpnService::class.java))
      } else if (pendingVpnStart) {
        Toast.makeText(this, R.string.permission_denied_vpn, Toast.LENGTH_LONG).show()
      }
      pendingVpnStart = false
    }

  private val notificationPermissionLauncher =
    registerForActivityResult(ActivityResultContracts.RequestPermission()) {
      if (pendingNotificationPermissionForVpnStart) {
        pendingNotificationPermissionForVpnStart = false
        continueVpnStartAfterNotificationCheck()
      }
    }

  override fun attachBaseContext(newBase: Context) {
    super.attachBaseContext(LocaleManager.localizedContext(newBase))
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    FileManager.initialize(applicationContext)
    ThemeHelper.getInstance().init()
    FontSizeHelper.init()
    applySystemBars(ThemeHelper.currentTheme)

    setContent {
      val theme by ThemeHelper.themeFlow.collectAsStateWithLifecycle()
      LaunchedEffect(theme) {
        applySystemBars(theme)
      }
      OblivionTheme {
        OblivionApp(
          onRequestVpnStart = { requestVpnStart() },
          onStopVpn = { OblivionVpnService.stopVpnService(this) },
        )
      }
    }
  }

  private fun requestVpnStart() {
    val core = VpnCoreType.getCurrent()
    if (!core.isReady) {
      Toast.makeText(this, core.availabilityReasonRes, Toast.LENGTH_LONG).show()
      return
    }
    if (shouldRequestNotificationPermissionBeforeVpnStart()) {
      FileManager.set(NOTIFICATION_PERMISSION_REQUESTED_KEY, true)
      pendingNotificationPermissionForVpnStart = true
      notificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
      return
    }

    continueVpnStartAfterNotificationCheck()
  }

  private fun shouldRequestNotificationPermissionBeforeVpnStart(): Boolean =
    Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
      ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) !=
        PackageManager.PERMISSION_GRANTED &&
      !FileManager.getBoolean(NOTIFICATION_PERMISSION_REQUESTED_KEY, false)

  private fun continueVpnStartAfterNotificationCheck() {
    val vpnIntent = VpnService.prepare(this)
    if (vpnIntent != null) {
      pendingVpnStart = true
      vpnPermissionLauncher.launch(vpnIntent)
    } else {
      startVpnServiceFromActivity(Intent(this, OblivionVpnService::class.java))
    }
  }

  private fun startVpnServiceFromActivity(intent: Intent) {
    startVpnService(this, intent)
    vpnServiceConnector.ensureBound()
  }

  private fun applySystemBars(theme: ThemeHelper.Theme) {
    val isLight = theme == ThemeHelper.Theme.LIGHT
    val barColor =
      when (theme) {
        ThemeHelper.Theme.OLED -> AndroidColor.BLACK
        ThemeHelper.Theme.DARK -> AndroidColor.rgb(0x0f, 0x0a, 0x00)
        ThemeHelper.Theme.LIGHT -> AndroidColor.rgb(0xff, 0xfb, 0xf4)
      }

    enableEdgeToEdge(
      statusBarStyle =
        if (isLight) {
          SystemBarStyle.light(barColor, barColor)
        } else {
          SystemBarStyle.dark(barColor)
        },
      navigationBarStyle =
        if (isLight) {
          SystemBarStyle.light(barColor, barColor)
        } else {
          SystemBarStyle.dark(barColor)
        },
    )
  }

  companion object {
    private const val NOTIFICATION_PERMISSION_REQUESTED_KEY = "notification_permission_requested"

    @JvmStatic
    fun start(context: Context) {
      val starter =
        Intent(context, MainActivity::class.java).apply {
          putExtra("origin", context.javaClass.simpleName)
          addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP)
        }
      context.startActivity(starter)
    }

    @JvmStatic
    fun startVpnService(context: Context, intent: Intent) {
      val dns = FileManager.getString(FileManager.Keys.USERSETTING_DNS).ifEmpty { "1.1.1.1" }

      intent.apply {
        putExtra(
          FileManager.Keys.USERSETTING_PROXYMODE,
          FileManager.getBoolean(FileManager.Keys.USERSETTING_PROXYMODE),
        )
        putExtra(
          FileManager.Keys.USERSETTING_ENDPOINT,
          FileManager.getString(FileManager.Keys.USERSETTING_ENDPOINT),
        )
        putExtra(FileManager.Keys.USERSETTING_DNS, dns)
        putExtra(
          FileManager.Keys.USERSETTING_VPN_DNS,
          FileManager.getString(FileManager.Keys.USERSETTING_VPN_DNS, dns),
        )
        action = OblivionVpnService.FLAG_VPN_START
      }
      ContextCompat.startForegroundService(context, intent)
    }
  }
}
